### ISSUE TITLE
Support dtype conversions on stick-unaligned tensors via padding

### ISSUE THREAD

**Issue by yoheiueda** [torch-spyre/torch-spyre#3999]
### What feature would you like to be added?

Support dtype conversions on tensors whose stick dimension is not a whole multiple of the stick width, by padding the stick dimension of the tensor that is short.

A stick is a fixed 128 bytes, so how many elements fit in one stick depends on the width of the dtype:

| dtype | bytes per element | elements per stick |
|---|---|---|
| `float32`, `int32` | 4 | 32 |
| `float16`, `bfloat16` | 2 | 64 |
| `float8_e4m3fn` | 1 | 128 |

A conversion does not change how many elements a tensor holds, but it does change how many bytes each element takes, so the same logical extent occupies a different number of sticks on each side of the conversion. Because the hardware reads and writes whole sticks, a conversion between two dtypes of different widths has to iterate over the coarser of the two stick grids. When the extent of the stick dimension does not divide evenly into that coarser grid, one side of the conversion has a stick that the tensor does not cover, and that side needs a padding stick so the access stays inside real storage.

The main case is `float16` ↔ `float32`, where the two stick widths are 64 and 32 elements. The conversion iterates the 64-element grid, so any last-dimension extent that is not a multiple of 64 is affected. Conversions involving 1-byte dtypes such as `float8_e4m3fn` are the same requirement at a different pair of widths, where the coarser grid is 128 elements.

### What the two directions look like

The two directions are the same missing stick on the same wide-dtype buffer. Widening writes a stick that the buffer does not own; narrowing reads a stick that nothing ever wrote. One padding stick on the wide side covers both.

Widening `float16` to `float32`, shape `(4, 96)` — the short side is the `float32` **output**:

<img width="1040" height="640" alt="Image" src="https://github.com/user-attachments/assets/a70fd6cd-357e-4114-83e1-cbaa81fe1630" />

Narrowing `float32` to `float16`, same shape — the short side is the `float32` **input**, and the resulting damage appears in the `float16` output:

<img width="1040" height="640" alt="Image" src="https://github.com/user-attachments/assets/cce4c110-1cb0-4f05-901a-1b3f4589a8aa" />

Cell widths in both figures are drawn to scale in bytes, so every stick is the same physical width and the dtype change appears as width: 96 elements occupy 192 bytes at `float16` (one and a half sticks) and 384 bytes at `float32` (three sticks). The conversion iterates 128 lanes, which is four `float32` sticks, and the `float32` buffer covers three.

The figures also show why the effect is not confined to a tail. One `float16` stick is formed by interleaving two consecutive `float32` sticks in groups of four elements. When the number of `float32` sticks is odd, the last `float16` stick has no partner stick holding live data, so the unbacked lanes are distributed through the stick rather than sitting after the last valid element.

### What has to change

Two things have to move together, and only one of them is allocation:

1. The **allocated size** of the wide-dtype tensor has to include the padding stick, so the access has real storage behind it.
2. The **declared extent** of that tensor has to include it as well, because every op that consumes the tensor derives its own iteration window from that extent. If only the allocation grows, a consumer keeps iterating the shorter window and the narrowing conversion still sources lanes that were never written.

The padded lanes only have to be addressable. Nothing accumulates across them, so unlike matmul padding they do not need to be zero-filled.

### Why is this feature needed?

Stick-unaligned dtype conversions are not currently supported, and today the unsupported case is not rejected — it compiles and returns wrong values for the affected shapes, with no diagnostic. A model that converts a tensor whose last dimension is not a multiple of 64 gets silently incorrect results.

The shapes involved are ordinary rather than exotic. Any hidden size, vocabulary slice, or head dimension that is a multiple of 32 but not 64 lands in the affected class, and mixed-precision models convert between `float16` and `float32` routinely. Quantized paths using 1-byte dtypes hit the same rule with a coarser grid, so more extents are affected there.

The existing tests do not cover it. The shapes used for dtype round-trip coverage are all multiples of 64, so none of them falls in the affected class, and a bare `x.to(float32).to(float16)` round-trip is folded away into an identity before it reaches the device, so it never exercises a real conversion at all. A regression test for this feature needs both an extent that is a multiple of 32 but not 64 and a genuine consumer of the wide-dtype value between the two conversions, so that the conversion is actually emitted.

Worth noting that the gap is not in SDSC generation. The iteration space is already rounded up correctly to the coarser stick grid, in both directions. What lags behind is the declared extent of the tensor between the conversions, and therefore the iteration window of everything that reads it.

### Any alternative solution that you may want to share?

Not really. Because the hardware reads and writes whole sticks, a conversion between dtypes of different widths necessarily touches a stick that a stick-unaligned tensor does not cover, and giving that access real storage is what padding means. The alternative to padding is to keep rejecting these shapes, which is what the feature is asking to remove.

There is a genuine choice about *where in the pipeline* the padding is applied, and it has real consequences for how invasive the change is. That is described below.

### Anything else we need to know?

**Where the padding is applied is the main open question.** The difficulty is one of ordering. The padding has to go on the **stick dimension**, but which dimension is the stick dimension is not known until layouts are committed, which happens in the `insert_restickify` pass. Anything earlier is reasoning about a layout that has not been decided yet. That gives two directions.

*Pad before layouts are committed.* Padding is applied ahead of `insert_restickify`, which means the standard stick layout has to be enforced for the tensors taking part in a conversion, so that the dimension being padded is known to be the stick dimension. This is workable and needs no new pass, but it constrains layout propagation and adds a special case to it, which makes propagation harder to reason about and easy to get subtly wrong. Best treated as a short-term solution.

*Pad after layouts are committed.* Padding is applied after `insert_restickify`, where the stick dimension is already known and nothing has to be assumed or enforced about layout. This is the cleaner formulation, since the pass works from decided facts instead of constraining an earlier decision, and it leaves layout propagation untouched. The cost is a new compiler pass. This is the better long-term solution.

The two are compatible: the first can unblock the common `float16` ↔ `float32` shapes while the second is built, provided the layout constraint it introduces is documented and kept confined to conversions.

**This is a different axis from restickify padding (#3110).** That work pads **non-stick** device dimensions, where the stick width is fixed on both sides and what moves is a dimension entering or leaving stick position under a transpose. Here the stick width itself changes, and the dimension that needs padding is the stick dimension.

| | restickify padding (#3110) | this feature |
|---|---|---|
| Trigger | a transpose moves a dimension in or out of stick position | the two dtypes have different stick widths |
| Stick width | fixed on both sides | changes between the sides |
| Dimension padded | a non-stick device dimension | the stick dimension |
| Round-up target | the dimension's own dtype stick | the **coarser** of the two dtype sticks |

That last row matters for reuse: rounding a tensor up against its own dtype's stick is a no-op here, because the extents in question are already multiples of the narrower grid. The round-up has to be taken against the other side's dtype.

### Related issues:
- #2214

**Comment by michihirohorie** [comment]
Repro code:
```
import torch

torch.set_printoptions(threshold=torch.inf)

DEVICE = torch.device("spyre")
torch.manual_seed(0xAFFE)

x = torch.ones(2, 96, dtype=torch.float16)

cpu_result = torch.add(x.float(), 0.0).to(torch.float16)
print(f"\nCPU result: {cpu_result.shape}")
print(cpu_result)

x_device = x.to(DEVICE)

compiled_fn = torch.compile(lambda a: torch.add(a.float(), 0.0).to(torch.float16))
compiled_result = compiled_fn(x_device).cpu()

print(f"\nSpyre Compiled result: {compiled_result.shape}")
print(compiled_result)

cpu_delta = torch.abs(compiled_result - cpu_result).max()
print(f"\nMax delta Compiled Spyre vs. CPU: {cpu_delta}")
```

Result:
The third stick seems be converted back to fp16 while still in the staggered layout. It looks as if the backend was unable to process fp32 sticks 3 and 4 together during the FP32→FP16 conversion, and instead simply emits the contents of stick 3 directly.

```
CPU result: torch.Size([2, 96])
tensor([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1.]], dtype=torch.float16)

Spyre Compiled result: torch.Size([2, 96])
tensor([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 0.,
         1., 1., 1., 1., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 1., 1.,
         1., 1., 0., 0., 0., 0.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0.,
         0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0.,
         1., 1., 1., 1., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 1., 1.,
         1., 1., 0., 0., 0., 0.]], dtype=torch.float16)

Max delta Compiled Spyre vs. CPU: 1.0
```

<details>
<summary>OpSpecs</summary>

```
# Topologically Sorted Source Nodes: [float_1, add, to], Original ATen: [aten._to_copy, aten.add]
# Source node to ATen node mapping:
#   add => add
#   float_1 => convert_element_type
#   to => convert_element_type_1
# Graph fragment:
#   %arg0_1 : Tensor "f16[2, 96][96, 1]spyre:0" = PlaceHolder[target=arg0_1]
#   %convert_element_type : Tensor "f32[2, 96][96, 1]spyre:0" = PlaceHolder[target=convert_element_type]
#   %constant_default : Tensor "f32[][]spyre:0" = PlaceHolder[target=constant_default]
#   %add : Tensor "f32[2, 96][96, 1]spyre:0" = PlaceHolder[target=add]
#   %convert_element_type : Tensor "f32[2, 96][96, 1]spyre:0"[num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%arg0_1, torch.float32), kwargs = {})
#   %add : Tensor "f32[2, 96][96, 1]spyre:0"[num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%convert_element_type, 0.0), kwargs = {})
#   %convert_element_type_1 : Tensor "f16[2, 96][96, 1]spyre:0"[num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%add, torch.float16), kwargs = {})
#   return %convert_element_type,%add,%convert_element_type_1
sdsc_fused__to_copy_add_0 = async_compile.sdsc('sdsc_fused__to_copy_add_0',
    [
        OpSpec(
            op='dl16tofp32',
            is_reduction=False,
            iteration_space={sympify('c0'): (sympify('2'), 2), sympify('c1'): (sympify('96'), 2)},
            op_info={},
            core_id_to_work_slice={sympify('c0'): sympify('Mod(core_id, 2)'), sympify('c1'): sympify('Mod(floor(core_id/2), 2)')},
            symbolic_dim_bounds={},
            debug_handle=DebugHandle(id=3778950086994218003, source=SourceLoc(file='/home/horie/dt-inductor/torch-spyre/examples/substick_fp32.py', start_line=16, start_col=0, end_line=None, end_col=None), aten_op='aten._to_copy.default', ir_chain=('convert_element_type', 'buf0'), fused_from=(), transform_history=()),
            args=[
                TensorArg(
                    is_input=True, arg_index=0, device_dtype=DataFormats.SEN169_FP16,
                    device_size=[2, 2, 64],
                    device_coordinates=[sympify('floor(c1/64)'), sympify('c0'), sympify('Mod(c1, 64)')],
                    allocation={'hbm': 0},
                ),
                TensorArg(
                    is_input=False, arg_index=-1, device_dtype=DataFormats.IEEE_FP32,
                    device_size=[4, 2, 32],
                    device_coordinates=[sympify('floor(c1/32)'), sympify('c0'), sympify('Mod(c1, 32)')],
                    allocation={'hbm_pool': 0},
                    element_arrangement=ElementArrangement.DL16_TO_FP32,
                ),
            ]
        ),
        OpSpec(
            op='add',
            is_reduction=False,
            iteration_space={sympify('c0'): (sympify('2'), 2), sympify('c1'): (sympify('96'), 3)},
            op_info={},
            core_id_to_work_slice={sympify('c0'): sympify('Mod(core_id, 2)'), sympify('c1'): sympify('Mod(floor(core_id/2), 3)')},
            symbolic_dim_bounds={},
            debug_handle=DebugHandle(id=4317739794579618317, source=SourceLoc(file='/home/horie/dt-inductor/torch-spyre/examples/substick_fp32.py', start_line=16, start_col=0, end_line=None, end_col=None), aten_op='aten.add.Tensor', ir_chain=('add', 'buf1'), fused_from=(), transform_history=(ProvenanceTransform(kind='rewrite', pass_name='split_multi_ops', reason='rewrite original buffer body'),)),
            args=[
                TensorArg(
                    is_input=True, arg_index=-1, device_dtype=DataFormats.IEEE_FP32,
                    device_size=[4, 2, 32],
                    device_coordinates=[sympify('floor(c1/32)'), sympify('c0'), sympify('Mod(c1, 32)')],
                    allocation={'hbm_pool': 0},
                    element_arrangement=ElementArrangement.DL16_TO_FP32,
                ),
                TensorArg(
                    is_input=True, arg_index=1, device_dtype=DataFormats.IEEE_FP32,
                    device_size=[1, 1, 32],
                    device_coordinates=[sympify('0'), sympify('0'), sympify('0')],
                    allocation={'hbm': 1},
                ),
                TensorArg(
                    is_input=False, arg_index=-1, device_dtype=DataFormats.IEEE_FP32,
                    device_size=[4, 2, 32],
                    device_coordinates=[sympify('floor(c1/32)'), sympify('c0'), sympify('Mod(c1, 32)')],
                    allocation={'hbm_pool': 1024},
                    element_arrangement=ElementArrangement.DL16_TO_FP32,
                ),
            ]
        ),
        OpSpec(
            op='fp32todl16',
            is_reduction=False,
            iteration_space={sympify('c0'): (sympify('2'), 2), sympify('c1'): (sympify('96'), 1)},
            op_info={},
            core_id_to_work_slice={sympify('c0'): sympify('Mod(core_id, 2)'), sympify('c1'): sympify('0')},
            symbolic_dim_bounds={},
            debug_handle=DebugHandle(id=8750251953419466956, source=SourceLoc(file='/home/horie/dt-inductor/torch-spyre/examples/substick_fp32.py', start_line=16, start_col=0, end_line=None, end_col=None), aten_op='aten._to_copy.default', ir_chain=('convert_element_type_1', 'buf2'), fused_from=(), transform_history=()),
            args=[
                TensorArg(
                    is_input=True, arg_index=-1, device_dtype=DataFormats.IEEE_FP32,
                    device_size=[4, 2, 32],
                    device_coordinates=[sympify('floor(c1/32)'), sympify('c0'), sympify('Mod(c1, 32)')],
                    allocation={'hbm_pool': 1024},
                    element_arrangement=ElementArrangement.DL16_TO_FP32,
                ),
                TensorArg(
                    is_input=False, arg_index=2, device_dtype=DataFormats.SEN169_FP16,
                    device_size=[2, 2, 64],
                    device_coordinates=[sympify('floor(c1/64)'), sympify('c0'), sympify('Mod(c1, 64)')],
                    allocation={'hbm': 2},
                ),
            ]
        ),
    ]
, pool_size=2048)


async_compile.wait(globals())
del async_compile

class Runner:
    def __init__(self, partitions):
        self.partitions = partitions

    def recursively_apply_fns(self, fns):
        new_callables = []
        for fn, c in zip(fns, self.partitions):
            new_callables.append(fn(c))
        self.partitions = new_callables

    def call(self, args):
        arg0_1, = args
        args.clear()
        buf3 = spyre_constant_tensor(0.0, torch.device("spyre:0"), torch.float32)
        assert_size_stride(arg0_1, (2, 96), (96, 1), 'input')
        buf2 = spyre_empty_with_layout((2, 96), (96, 1), torch.float16, SpyreTensorLayout(device_size=[2, 2, 64], stride_map =[64, 96, 1], device_dtype=DataFormats.SEN169_FP16))
        # [Provenance debug handles] sdsc_fused__to_copy_add_0:1
        sdsc_fused__to_copy_add_0.run(arg0_1, buf3, buf2)
        del arg0_1
        return (buf2, )

runner = Runner(partitions=[])
call = runner.call
recursively_apply_fns = runner.recursively_apply_fns
```

</details>

[sdsc_fp32_substick.zip](https://github.com/user-attachments/files/32086475/sdsc_fp32_substick.zip)
